src/main/java/bio/terra/service/snapshot/SnapshotService.java

```diff
@@ -359,6 +361,29 @@ public UUID getFirstSourceDatasetIdFromSnapshotId(
         .orElseThrow(() -> new DatasetNotFoundException("Source dataset for snapshot not found"));
   }
 
+  public SnapshotPreviewModel retrievePreview(UUID snapshotId, String tableName, int count) {
```

Comment:
One thought; I was thinking it would be nice to not have the service layer code know about the model, if possible, since the model is part of the controller layer. The only difference here would be that this code would return the `List<Map<>>` and the caller would need to wrap it in the model object. Conceptually this seems correct to me; that the different layers shouldn't have to know about each other, but I haven't looked at it closely enough to know what the cost would be. It's possible it doesn't make sense in cases where the request model is complex.

Reply:
I'm not sure about this one, most of the other items in the `SnapshotService` return `Model`s of some kind. I'm open to switching it, but I did it to be consistent with the other methods.